caffe2/operators/onnx_while_op.h

```diff
@@ -0,0 +1,224 @@
+  // Body
+  //  Inputs: iteration number, condition, loop-carried dependencies
+  //  Outputs: condition, loop-carried dependencies, scan_outputs
+  bool RunOnDevice() override {
+    constexpr int64_t num_inputs_before_lcds = 2;
+    // First input is the maximumt trip count. Second input is the condition
+    // variable (for the first iteration). The rest of the inputs are
+    // loop-carried dependencies.
+    int num_loop_carried_deps = InputSize() - num_inputs_before_lcds;
+    int64_t max_trip_count = *Input(0).template data<int64_t>();
+    const int64_t first_iter_condition = *Input(1).template data<int64_t>();
+
+    // Body graph has 2+N inputs: iteration number, condition value, and N
+    // loop-carried dependencies
+    CAFFE_ENFORCE_EQ(
+        num_loop_carried_deps + 2, body_net_->external_input().size(),
+        "Body graph must have 2+N inputs, where N is the number of "
+        "loop carried dependencies.");
+
+    // Body graph has 1+N+K outputs: recalculated condition variable, N
+    // loop-carried dependencies, and K scan_outputs
+    int num_scan_outputs =
+        body_net_->external_output().size() - num_loop_carried_deps - 1;
+
+    CAFFE_ENFORCE_GE(num_scan_outputs, 0,
+                     "Body graph must have N+K outputs, where N is the number "
+                     "of loop-carried dependencies and K is the number of scan "
+                     "outputs");
+
+    // Copy initial loop-carried dependencies
+    for (int i = 0; i < num_loop_carried_deps; ++i) {
+      lcd_tensors_[i]->CopyFrom(Input(i + num_inputs_before_lcds));
+    }
+
+    // Initialize iteration variable
+    iteration_var_->Resize(1);
+    auto *iteration_var_ptr = iteration_var_->template mutable_data<int64_t>();
+    *iteration_var_ptr = 0ll;
+
+    // Input condition var. This requires special handling
+    input_condition_var_->Resize(1);
+    auto *input_condition_var_ptr =
+        input_condition_var_->template mutable_data<int64_t>();
+    *input_condition_var_ptr = first_iter_condition;
+
+    // Output condition var. This is yielded by the body net and we will use its
+    // value to determine further iteration
+
+    condition_var_->Resize(1);
+    auto *condition_var_ptr = condition_var_->template mutable_data<int64_t>();
+
+    auto valid_iter_num = [this, max_trip_count](int64_t i) {
+      if (has_trip_count_) {
+        return i < max_trip_count;
+      } else {
+        return true;
+      }
+    };
+
+    auto condition_true = [this, first_iter_condition,
+                           condition_var_ptr](int64_t i) {
+      if (has_cond_) {
+        if (i == 0) {
+          return (bool)first_iter_condition;
+        } else {
+          return (bool)*condition_var_ptr;
+        }
+      } else {
+        return true;
+      }
+    };
+
+    // Allocate scan_outputs for zero-iteration case
+    for (int i = 0; i < num_scan_outputs; ++i) {
+      Output(i + num_loop_carried_deps)->Resize(0);
```

Comment:
isn't there a number of iterations first according to the comment above? (then lcd, then scan outputs)

Reply:
Good catch, that comment above is stale and left-over from before a recent spec change. I'll update it